controllers/openshift_controller.go

```diff
@@ -2016,6 +2032,94 @@ func (r *KataConfigOpenShiftReconciler) clearUninstallStatus() {
 	r.kataConfig.Status.UnInstallationStatus.Failed.FailedNodesCount = 0
 }
 
+func (r *KataConfigOpenShiftReconciler) findInProgressCondition() *kataconfigurationv1.KataConfigCondition {
+	for i := 0; i < len(r.kataConfig.Status.Conditions); i++ {
+		if r.kataConfig.Status.Conditions[i].Type == kataconfigurationv1.KataConfigInProgress {
+			return &r.kataConfig.Status.Conditions[i]
+		}
+	}
+	return nil
+}
+
+func (r *KataConfigOpenShiftReconciler) getInProgressConditionValue() corev1.ConditionStatus {
+	cond := r.findInProgressCondition()
+	if cond == nil {
+		return corev1.ConditionUnknown
+	}
+	return cond.Status
+}
+
+func (r *KataConfigOpenShiftReconciler) addInProgressCondition() *kataconfigurationv1.KataConfigCondition {
+	r.kataConfig.Status.Conditions = append(r.kataConfig.Status.Conditions, kataconfigurationv1.KataConfigCondition{Type: kataconfigurationv1.KataConfigInProgress})
+
+	r.Log.Info("InProgress Condition added")
+
+	return &r.kataConfig.Status.Conditions[len(r.kataConfig.Status.Conditions)-1]
+}
+
+// This is just a technical helper to all InProgress Condition mutators,
+// factoring their common preamble out into an own function.
+func (r *KataConfigOpenShiftReconciler) retrieveInProgressConditionForChange() *kataconfigurationv1.KataConfigCondition {
+	cond := r.findInProgressCondition()
+	if cond == nil {
+		cond = r.addInProgressCondition()
+	}
+
+	cond.LastTransitionTime = metav1.Now()
+
+	return cond
+}
+
+func (r *KataConfigOpenShiftReconciler) setInProgressConditionToInstalling() {
+	cond := r.retrieveInProgressConditionForChange()
+	cond.Status = corev1.ConditionTrue
+	cond.Reason = "Installing"
+	cond.Message = "Performing initial installation on cluster"
+
+	r.Log.Info("InProgress Condition set to Installing")
+}
+
+func (r *KataConfigOpenShiftReconciler) setInProgressConditionToUninstalling() {
+	cond := r.retrieveInProgressConditionForChange()
+	cond.Status = corev1.ConditionTrue
+	cond.Reason = "Uninstalling"
+	cond.Message = "Removing kata from cluster"
+
+	r.Log.Info("InProgress Condition set to Uninstalling")
+}
+
+func (r *KataConfigOpenShiftReconciler) setInProgressConditionToUpdating() {
+	cond := r.retrieveInProgressConditionForChange()
+	cond.Status = corev1.ConditionTrue
+	cond.Reason = "Updating"
+	cond.Message = "Adding and/or removing kata-enabled nodes"
+
+	r.Log.Info("InProgress Condition set to Updating")
+}
+
+func (r *KataConfigOpenShiftReconciler) setInProgressConditionToFailed(failingNode *corev1.Node) {
+	reasonForDegraded, ok := failingNode.Annotations["machineconfiguration.openshift.io/reason"]
+	if !ok {
+		r.Log.Info("Missing machineconfiguration.openshift.io/reason on Degraded node", "node", failingNode.GetName())
+	}
+
+	cond := r.retrieveInProgressConditionForChange()
+	cond.Status = corev1.ConditionTrue
+	cond.Reason = "Failed"
+	cond.Message = "Node " + failingNode.GetName() + " Degraded: " + reasonForDegraded
+
```

Comment:
Mention kata as well here ?

Reply:
Not quite sure how though.

Reply:
> Not quite sure how though.

Callers of `setInProgressConditionToFailed()` know if install or uninstall is in progress. This could be passed as a bool and the message would be "Node XXX got degraded with reason YYY while  (un)installing kata".

Reply:
Ah okay, that's true.  Although the information whether the affected Node was installing or uninstalling is readily available by looking at `kataNodes.failedToInstall` and `kataNodes.failedToUninstall` lists.

Generally, a Condition.message is a simple string and, I believe, not designed or meant to carry high-resolution information.  Consider an actually frequently happening case of multiple Nodes becoming Degraded at the same time.  Condition.message can never do justice to accurately describing this.  Sometimes Nodes can even become Degraded while installing kata as others while uninstalling it, possibly with different `machineconfiguration.openshift.io/reason`s.

The idea of InProgress Condition.message seems to me to be more like "something went wrong, please check out `kataNodes.failedToInstall` and `kataNodes.failedToUninstall` for more detail, or `Node.metadata.annotations.machineconfiguration.openshift.io/reason` for affected Nodes if even that is not enough".

Maybe the current form is misleading, trying to save some work to the user by including the MCO `reason` message for the simple and probably most common cases of a single Node becoming Degraded or a group of Nodes becoming Degraded for the same reason, and thus raising false expectations that does more than it actually can?

Reply:
Come to think about it, how about we include each failed Node's `machineconfigmachineconfiguration.openshift.io/reason` in parens after Node name directly the `failedToInstall` and `failedToUninstall` lists and restrict the `Condition.message` to just something along the lines of "Node(s) failed, please see the failedToInstall and failedToUninstall lists for more information"?